File: hmvvm/src/main/java/com/see99/hmvvm/base/BaseActivity.java
package com.see99.hmvvm.base;

import android.content.Context;
import android.os.Bundle;
import android.view.View;
import android.view.inputmethod.InputMethodManager;

import androidx.annotation.LayoutRes;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import androidx.databinding.DataBindingUtil;
import androidx.databinding.ViewDataBinding;
import androidx.lifecycle.Observer;

import com.see99.hmvvm.MyApp;
import com.see99.hmvvm.dagger.component.ActivityComponent;
import com.see99.hmvvm.dagger.component.DaggerActivityComponent;

import javax.inject.Inject;


public abstract class BaseActivity<T extends ViewDataBinding, V extends BaseViewModel> extends AppCompatActivity {

    protected T binding;

    @Inject
    protected V mViewModel;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        performInject(getActivityComponent());
        super.onCreate(savedInstanceState);
        performDataBinding();
        initBaseObserver();
    }

    private void initBaseObserver() {
        mViewModel.routers.observe(this, new Observer<Router>() {
            @Override
            public void onChanged(Router router) {
                if (router.getBundle()!=null){
                    performRouter(router.getRouterId(),router.getBundle());
                }else {
                    performRouter(router.getRouterId());
                }
            }
        });
    }

    /**
     * 跳转的方式
     * 带参数跳转
     * 无参数跳转
     *
     * 页面跳转，调试路由传递的参数
     */
    protected  void performRouter(int routerId,Bundle bundle){

    }

    /**
     * 无参数跳转
     * @param routerId
     */
    public void performRouter(int routerId){

    }

    //注入方法
    protected abstract void performInject(ActivityComponent activityComponent);

    private ActivityComponent getActivityComponent() {
        return DaggerActivityComponent.builder()
                .setAppComponent(((MyApp)getApplication()).appComponent)
                .activity(this).build();
    }

    private void performDataBinding() {
        binding = DataBindingUtil.setContentView(this,getLayoutId());
        binding.setVariable(getVariableId(),mViewModel);
        binding.setLifecycleOwner(this);
        binding.executePendingBindings();
    }

    //绑定数据的id
    protected abstract int getVariableId();

    @LayoutRes
    public abstract int getLayoutId();

    //隐藏软键盘
    public void hideKeyboard() {
        View view = this.getCurrentFocus();
        if (view != null) {
            InputMethodManager imm = (InputMethodManager) getSystemService(Context.INPUT_METHOD_SERVICE);
            if (imm != null) {
                imm.hideSoftInputFromWindow(view.getWindowToken(), 0);
            }
        }
    }
}

File: hmvvmkotlin/src/main/java/com/see99/hmvvmkotlin/base/BaseActivity.kt
package com.see99.hmvvmkotlin.base

import android.os.Bundle
import androidx.annotation.LayoutRes
import androidx.appcompat.app.AppCompatActivity
import androidx.databinding.DataBindingUtil
import androidx.databinding.ViewDataBinding
import com.see99.hmvvmkotlin.MyApp
import com.see99.hmvvmkotlin.R
import com.see99.hmvvmkotlin.dagger.component.ActivityComponent
import com.see99.hmvvmkotlin.dagger.component.DaggerActivityComponent
import com.see99.utilslib.StatusBarUtils
import javax.inject.Inject

abstract class BaseActivity<T : ViewDataBinding, V : BaseViewModel> : AppCompatActivity() {

    @Inject
    lateinit var viewModel: V
    lateinit var binding: T


    override fun onCreate(savedInstanceState: Bundle?) {
        performInject(activityComponent())
        super.onCreate(savedInstanceState)
        StatusBarUtils.setStatusBarColor(this, R.color.statusbar)
        performDataBinding()
        initBaseObserver()

        initObserver()
        initView()
    }

    abstract fun initView()

    abstract fun initObserver()

    private fun initBaseObserver() {
        viewModel.routers.observe(this) {
            if (it.bundle != null) {
                performRouter(it.routerId, it.bundle!!)
            } else {
                performRouter(it.routerId)
            }
        }

    }

    /**
     * 当前页可以跳转到多个页   可以根据routerId  进行判断   默认routerId 为 0
     */
    open fun performRouter(routerId: Int, bundle: Bundle) {
    }

    /**
     * 跳转页面不用传递参数
     */
    open fun performRouter(routerId: Int) {
        if (routerId == 0){
            performRouter()
        }
    }

    /**
     * routerid 默认为 0
     */
    open fun performRouter() {

    }



    private fun performDataBinding() {
        binding = DataBindingUtil.setContentView(this, getLayoutId())
        binding.setVariable(getVariableId(), viewModel)
        binding.lifecycleOwner = this
        binding.executePendingBindings()

    }

    abstract fun getVariableId(): Int

    @LayoutRes
    abstract fun getLayoutId(): Int

    /**
     * 注入方法
     */
    abstract fun performInject(activityComponent: ActivityComponent)

//        private fun activityComponent(): ActivityComponent = DaggerActivityComponent.builder()
//            .appComponent((application as MyApp).appComponent).build()

    private fun activityComponent():ActivityComponent{
      return  DaggerActivityComponent.builder().activity(this)
                .appComponent((application as MyApp).appComponent)
                .build()
    }

}